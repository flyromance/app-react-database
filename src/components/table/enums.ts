
export enum PagingPosition {
  Bottom = 'bottom',
  Top = 'top',
  TopAndBottom = 'topAndBottom',
  BottomRight = 'bottomRight',
  TopRight = 'top',
}

export enum DataType {
  Boolean = 'boolean',
  Date = 'date',
  Number = 'number',
  Object = 'object',
  String = 'string',
}

export enum UiType {
  Boolean = 'boolean',
  Date = 'date',
  Number = 'number',
  Object = 'object',
  String = 'string',
}

export enum EditingMode {
  None = 'none',
  Cell = 'cell',
}

export enum ActionType {
  ClearFocused = 'ClearFocused',
  ClearSingleAction = 'ClearSingleAction',
  CloseEditor = 'CloseEditor',
  CloseRowEditors = 'CloseRowEditors',
  DeleteRow = 'DeleteRow',
  DeselectAllFilteredRows = 'DeselectAllFilteredRows',
  DeselectAllRows = 'DeselectAllRows',
  DeselectAllVisibleRows = 'DeselectAllVisibleRows',
  DeselectRow = 'DeselectRow',
  HideColumn = 'HideColumn',
  HideDetailsRow = 'HideDetailsRow',
  HideLoading = 'HideLoading',
  HideNewRow = 'HideNewRow',
  LoadData = 'LoadData',
  MoveFocusedDown = 'MoveFocusedDown',
  MoveFocusedLeft = 'MoveFocusedLeft',
  MoveFocusedRight = 'MoveFocusedRight',
  MoveFocusedUp = 'MoveFocusedUp',
  OpenEditor = 'OpenEditor',
  OpenRowEditors = 'OpenRowEditors',
  ReorderColumns = 'ReorderColumns',
  ReorderRows = 'ReorderRows',
  ResizeColumn = 'ResizeColumn',
  SaveNewRow = 'SaveNewRow',
  SaveRowEditors = 'SaveRowEditors',
  ScrollTable = 'ScrollTable',
  Search = 'Search',
  SelectAllFilteredRows = 'SelectAllFilteredRows',
  SelectAllRows = 'SelectAllRows',
  SelectAllVisibleRows = 'SelectAllVisibleRows',
  SelectRow = 'SelectRow',
  SelectRowsRange = 'SelectRowsRange',
  SelectSingleRow = 'SelectSingleRow',
  SetFocused = 'SetFocused',
  SetSingleAction = 'SetSingleAction',
  ShowColumn = 'ShowColumn',
  ShowDetailsRow = 'ShowDetailsRow',
  ShowLoading = 'ShowLoading',
  ShowNewRow = 'ShowNewRow',
  UpdateCellValue = 'UpdateCellValue',
  UpdateData = 'UpdateData',
  UpdateEditorValue = 'UpdateEditorValue',
  UpdateFilterRowOperator = 'UpdateFilterRowOperator',
  UpdateFilterRowValue = 'UpdateFilterRowValue',
  UpdateGroupsExpanded = 'UpdateGroupsExpanded',
  UpdatePageIndex = 'UpdatePageIndex',
  UpdatePageSize = 'UpdatePageSize',
  UpdatePagesCount = 'UpdatePagesCount',
  UpdateRow = 'UpdateRow',
  UpdateSortDirection = 'UpdateSortDirection',
  UpdateVirtualScrolling = 'UpdateVirtualScrolling',
}

export enum KeyboardEnum {
  Esc = 27,
  Enter = 13,
}

export enum SortDirection {
  Ascend = 'ascend',
  Descend = 'descend',
}

export enum SortingMode {
  None = 'none',
  Single = 'single',
  SingleTripleState = 'singleTripleState',
  SingleRemote = 'singleRemote',
  SingleTripleStateRemote = 'singleTripleStateRemote',
  MultipleRemote = 'multipleRemote',
  MultipleTripleStateRemote = 'multipleTripleStateRemote',
}

export enum FilteringMode {
  None = 'none',
  FilterRow = 'filterRow',
}

export enum FilterOperatorName {
  Equal = '=',
  MoreThan = '>',
  LessThan = '<',
  MoreThanOrEqual = '>=',
  LessThanOrEqual = '<=',
  Contains = 'contains',
  IsEmpty = 'IsEmpty',
  IsNotEmpty = 'IsNotEmpty',
}
