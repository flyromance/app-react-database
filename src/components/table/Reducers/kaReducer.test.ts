import { ITableProps } from '../';
import {
  clearSingleAction, deleteRow, deselectAllFilteredRows, deselectAllRows, deselectAllVisibleRows,
  deselectRow, loadData, reorderColumns, reorderRows, selectAllFilteredRows, selectAllRows,
  selectAllVisibleRows, selectRowsRange, selectSingleRow, setSingleAction, updateData,
} from '../actionCreators';
import { ActionType, FilterOperatorName } from '../enums';
import { kaReducer } from './kaReducer';

describe('kaReducer', () => {
  describe('SelectRowsRange', () => {
    const intialState = {
      data: [
        { key: 1 },
        { key: 2 },
        { key: 3 },
        { key: 4 },
      ],
      columns: [{ key: 'key' }],
      rowKeyField: 'key',
    };
    it('default', () => {
      const result: ITableProps = kaReducer(intialState, selectRowsRange(2, 4));
      expect(result.selectedRows).toMatchSnapshot();
    });
    it('revert set', () => {
      const result: ITableProps = kaReducer(intialState, selectRowsRange(4, 2));
      expect(result.selectedRows).toMatchSnapshot();
    });
    it('rowKeyValueTo is not specified', () => {
      const result: ITableProps = kaReducer(intialState, selectRowsRange(2, null));
      expect(result.selectedRows).toMatchSnapshot();
    });
    it('skip if rowKeyValueTo is not found', () => {
      const result: ITableProps = kaReducer(intialState, selectRowsRange(2, 7));
      expect(result.selectedRows).toMatchSnapshot();
    });
    it('skip if rowKeyValue is not found', () => {
      const result: ITableProps = kaReducer(intialState, selectRowsRange(7, 2));
      expect(result.selectedRows).toMatchSnapshot();
    });
  });
  describe('ReorderColumns', () => {
    it('default', () => {
      const intialState = {
        columns: [
          { key: '1' },
          { key: '2' },
          { key: '3' },
          { key: '4' },
        ],
        rowKeyField: 'id',
      };
      const result = kaReducer(intialState, reorderColumns('1', '3'));
      expect(result).toMatchSnapshot();
    });
  });
  describe('ReorderRows', () => {
    it('default', () => {
      const intialState = {
        data: [
          { id: 1 },
          { id: 2 },
          { id: 3 },
          { id: 4 },
        ],
        rowKeyField: 'id',
      };
      const result = kaReducer(intialState, reorderRows(1, 3));
      expect(result).toMatchSnapshot();
    });
    it('inner keys', () => {
      const intialState = {
        data: [
          { a: { id: 1 } },
          { a: { id: 2 } },
          { a: { id: 3 } },
          { a: { id: 4 } },
        ],
        rowKeyField: 'a.id',
      };
      const result = kaReducer(intialState, reorderRows(1, 3));
      expect(result).toMatchSnapshot();
    });
  });
  it('ScrollTable', () => {
    const scrollTop = 10;
    const intialState = {
      columns: [],
      data: [],
      rowKeyField: '',
    };
    const newState = kaReducer(intialState, { type: ActionType.ScrollTable, scrollTop });
    expect(newState).toEqual({ ...intialState, virtualScrolling: { scrollTop } });
  });
  it('SelectAllRows', () => {
    const intialState = {
      columns: [],
      data: [{ id: 1 }, { id: 2 }],
      rowKeyField: 'id',
    };
    const newState = kaReducer(intialState, selectAllRows());
    expect(newState).toEqual({ ...intialState, selectedRows: [1, 2] });
  });
  describe('SelectAllFilteredRows', () => {
    it('filter only', () => {
      const intialState: ITableProps = {
        columns: [{ key: 'id', filterRowValue: 2, filterRowOperator: FilterOperatorName.LessThanOrEqual }],
        data: [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }],
        rowKeyField: 'id',
      };
      const newState = kaReducer(intialState, selectAllFilteredRows());
      expect(newState).toEqual({ ...intialState, selectedRows: [1, 2] });
    });
    it('search', () => {
      const intialState: ITableProps = {
        columns: [{ key: 'id' }],
        data: [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }],
        rowKeyField: 'id',
        searchText: '1',
      };
      const newState = kaReducer(intialState, selectAllFilteredRows());
      expect(newState).toEqual({ ...intialState, selectedRows: [1] });
    });
  });

  describe('SelectAllVisibleRows', () => {
    it('default', () => {
      const intialState: ITableProps = {
        columns: [{ key: 'id', filterRowValue: 3, filterRowOperator: FilterOperatorName.LessThanOrEqual }],
        data: [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }],
        rowKeyField: 'id',
        paging: {
          enabled: true,
          pageSize: 2
        }
      };
      const newState = kaReducer(intialState, selectAllVisibleRows());
      expect(newState).toEqual({ ...intialState, selectedRows: [1, 2] });
    });
  });
  describe('DeselectAllVisibleRows', () => {
    it('default', () => {
      const intialState: ITableProps = {
        columns: [{ key: 'id', filterRowValue: 4, filterRowOperator: FilterOperatorName.LessThanOrEqual }],
        data: [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }, { id: 5 }],
        selectedRows: [1, 2, 3, 5],
        rowKeyField: 'id',
        paging: {
          enabled: true,
          pageSize: 2
        }
      };
      const newState = kaReducer(intialState, deselectAllVisibleRows());
      expect(newState).toEqual({ ...intialState, selectedRows: [3, 5] });
    });
  });

  describe('DeselectAllFilteredRows', () => {
    it('filter only', () => {
      const intialState: ITableProps = {
        columns: [{ key: 'id', filterRowValue: 2, filterRowOperator: FilterOperatorName.LessThanOrEqual }],
        data: [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }],
        selectedRows: [1, 2, 3],
        rowKeyField: 'id',
      };
      const newState = kaReducer(intialState, deselectAllFilteredRows());
      expect(newState).toEqual({ ...intialState, selectedRows: [3] });
    });
    it('search', () => {
      const intialState: ITableProps = {
        columns: [{ key: 'id' }],
        data: [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }],
        rowKeyField: 'id',
        selectedRows: [1, 2, 3],
        searchText: '1',
      };
      const newState = kaReducer(intialState, deselectAllFilteredRows());
      expect(newState).toEqual({ ...intialState, selectedRows: [2, 3] });
    });
  });
  it('SelectSingleRow', () => {
    const intialState = {
      columns: [],
      data: [{ id: 1 }, { id: 2 }],
      rowKeyField: 'id',
      selectedRows: [1],
    };
    const newState = kaReducer(intialState, selectSingleRow(2));
    expect(newState).toEqual({ ...intialState, selectedRows: [2] });
  });
  it('DeleteRow', () => {
    const intialState = {
      columns: [],
      data: [{ id: 1 }, { id: 2 }],
      rowKeyField: 'id',
    };
    const newState = kaReducer(intialState, deleteRow(2));
    expect(newState).toEqual({ ...intialState, data: [{ id: 1 }] });
  });
  it('DeselectAllRows', () => {
    const intialState = {
      columns: [],
      data: [{ id: 1 }, { id: 2 }],
      rowKeyField: 'id',
      selectedRows: [1, 2],
    };
    const newState = kaReducer(intialState, deselectAllRows());
    expect(newState).toEqual({ ...intialState, selectedRows: [] });
  });
  it('DeselectRow', () => {
    const intialState = {
      columns: [],
      data: [{ id: 1 }, { id: 2 }],
      rowKeyField: 'id',
      selectedRows: [1, 2],
    };
    const newState = kaReducer(intialState, deselectRow(2));
    expect(newState).toEqual({ ...intialState, selectedRows: [1] });
  });
  it('UpdateData', () => {
    const newData =  [{ id: 3 }, { id: 4 }];
    const intialState = {
      columns: [],
      data: [{ id: 1 }, { id: 2 }],
      rowKeyField: 'id',
      selectedRows: [1, 2],
    };
    const newState = kaReducer(intialState, updateData(newData));
    expect(newState.data).toEqual(newData);
  });
  it('ShowLoading', () => {
    const intialState = {
      loading: {
        enabled: false,
        text: 'test'
      }
    };
    const newState = kaReducer(intialState, { type: ActionType.ShowLoading });
    expect(newState.loading.enabled).toEqual(true);
    expect(newState.loading.text).toEqual('test');
  });
  it('ShowLoading - change text', () => {
    const intialState = {
      loading: {
        enabled: false,
        text: 'test'
      }
    };
    const newState = kaReducer(intialState, { type: ActionType.ShowLoading, text: '2' });
    expect(newState.loading.enabled).toEqual(true);
    expect(newState.loading.text).toEqual('2');
  });
  it('HideLoading', () => {
    const intialState = {
      loading: {
        enabled: true,
        text: 'test'
      }
    };
    const newState = kaReducer(intialState, { type: ActionType.HideLoading });
    expect(newState.loading.enabled).toEqual(false);
    expect(newState.loading.text).toEqual('test');
  });
  it('HideLoading', () => {
    const intialState = {
      loading: {
        enabled: true,
        text: 'test'
      }
    };
    const newState = kaReducer(intialState, { type: ActionType.HideLoading });
    expect(newState.loading.enabled).toEqual(false);
    expect(newState.loading.text).toEqual('test');
  });
  it('UpdatePageIndex', () => {
    const intialState = {
      paging: {
        enabled: true,
        pageIndex: 2,
      }
    };
    const newState = kaReducer(intialState, { type: ActionType.UpdatePageIndex, pageIndex: 5 });
    expect(newState.paging.enabled).toEqual(true);
    expect(newState.paging.pageIndex).toEqual(5);
  });
  it('SetSingleAction', () => {
    const intialState = { };
    const newState = kaReducer(intialState, setSingleAction(loadData()));
    expect(newState.singleAction).toMatchSnapshot();
  });
  it('ClearSingleAction', () => {
    const intialState = {
      singleAction: loadData(),
    };
    const newState = kaReducer(intialState, clearSingleAction());
    expect(newState.singleAction).toBeUndefined();
  });
});
