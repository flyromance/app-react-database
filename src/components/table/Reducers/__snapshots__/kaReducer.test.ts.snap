// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kaReducer ReorderColumns default 1`] = `
Object {
  "columns": Array [
    Object {
      "key": "2",
    },
    Object {
      "key": "3",
    },
    Object {
      "key": "1",
    },
    Object {
      "key": "4",
    },
  ],
  "rowKeyField": "id",
}
`;

exports[`kaReducer ReorderRows default 1`] = `
Object {
  "data": Array [
    Object {
      "id": 2,
    },
    Object {
      "id": 3,
    },
    Object {
      "id": 1,
    },
    Object {
      "id": 4,
    },
  ],
  "rowKeyField": "id",
}
`;

exports[`kaReducer ReorderRows inner keys 1`] = `
Object {
  "data": Array [
    Object {
      "a": Object {
        "id": 2,
      },
    },
    Object {
      "a": Object {
        "id": 3,
      },
    },
    Object {
      "a": Object {
        "id": 1,
      },
    },
    Object {
      "a": Object {
        "id": 4,
      },
    },
  ],
  "rowKeyField": "a.id",
}
`;

exports[`kaReducer SelectRowsRange default 1`] = `
Array [
  2,
  3,
  4,
]
`;

exports[`kaReducer SelectRowsRange revert set 1`] = `
Array [
  2,
  3,
  4,
]
`;

exports[`kaReducer SelectRowsRange rowKeyValueTo is not specified 1`] = `
Array [
  2,
]
`;

exports[`kaReducer SelectRowsRange skip if rowKeyValue is not found 1`] = `
Array [
  undefined,
  1,
  2,
]
`;

exports[`kaReducer SelectRowsRange skip if rowKeyValueTo is not found 1`] = `
Array [
  undefined,
  1,
  2,
]
`;

exports[`kaReducer SetSingleAction 1`] = `
Object {
  "type": "LoadData",
}
`;
